src/analysis/normed_space/deriv.lean

```diff
@@ -0,0 +1,178 @@
+/-
+Copyright (c) 2019 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad
+
+The Fréchet derivative.
+
+Let `E` and `F` be normed spaces, and `f : E → F`. Then
+
+  `has_fderiv_at_within f f' x s`
+
+says that the function `f' : E → F` is a bounded linear map and `f` has derivative `f'` at
+`x`, where the domain of interest is restricted to `s`. We also have
+
+  `has_fderiv_at f f' x := has_fderiv_at_within f f' x univ`
+
+We may want to prove some theorems specifically for this special case, but most of the more
+general theorems should apply right out of the box.
+
+The derivative is characterized in two ways: in terms of the `tendsto` relation, and in terms of
+the `littleo` relation.
+-/
+import topology.basic analysis.normed_space.bounded_linear_maps ..asymptotics tactic.abel
+open filter asymptotics
+
+variables {E : Type*} [normed_space ℝ E]
+variables {F : Type*} [normed_space ℝ F]
+variables {G : Type*} [normed_space ℝ G]
+
+def has_fderiv_at_within (f : E → F) (f' : E → F) (x : E) (s : set E) :=
+is_bounded_linear_map f' ∧
+  tendsto (λ x', ∥x' - x∥⁻¹ • (f x' - f x - f' (x' - x))) (nhds_within x s) (nhds 0)
+
+def has_fderiv_at (f : E → F) (f' : E → F) (x : E) :=
+has_fderiv_at_within f f' x set.univ
+
+theorem has_fderiv_equiv_aux (f : E → F) (f' : E → F) (x : E) (s : set E)
+    (bf' : is_bounded_linear_map f') :
+  tendsto (λ x', ∥x' - x∥⁻¹ • (f x' - f x - f' (x' - x))) (nhds_within x s) (nhds 0) =
+  tendsto (λ x', ∥x' - x∥⁻¹ * ∥f x' - f x - f' (x' - x)∥) (nhds_within x s) (nhds 0) :=
+begin
+  have f'0 : f' 0 = 0 := bf'.to_linear_map.map_zero,
+  rw [tendsto_iff_norm_tendsto_zero], congr,
+  ext x',
+  have : ∥x' + -x∥⁻¹ ≥ 0, from inv_nonneg.mpr (norm_nonneg _),
+  simp [norm_smul, real.norm_eq_abs, abs_of_nonneg this]
+end
+
+theorem has_fderiv_iff_littleo {f : E → F} {f' : E → F} {x : E} {s : set E} :
+  has_fderiv_at_within f f' x s ↔
+    is_bounded_linear_map f' ∧
+      is_littleo (λ x', f x' - f x - f' (x' - x)) (λ x', x' - x) (nhds_within x s) :=
+and.congr_right_iff.mpr $
+  assume bf' : is_bounded_linear_map f',
+  have f'0 : f' 0 = 0 := bf'.to_linear_map.map_zero,
+  have h : ∀ x', ∥x' - x∥ = 0 → ∥f x' - f x - f' (x' - x)∥ = 0, from
+    assume x' hx',
+    have x' = x,
+      from eq_of_sub_eq_zero ((norm_eq_zero _).mp hx'),
+    by rw this; simp [f'0],
+  begin
+    rw has_fderiv_equiv_aux _ _ _ _ bf',
+    rw ←is_littleo_norm_left,
+    rw ←is_littleo_norm_right,
+    rw is_littleo_iff_tendsto h,
+    apply iff_of_eq, congr, ext x',
+    apply mul_comm
+  end
+
+def has_fderiv_at_within_mono {f : E → F} {f' : E → F} {x : E} {s t : set E}
+    (h : has_fderiv_at_within f f' x t) (hst : s ⊆ t) :
```

Comment:
These two arguments should be swapped, and it should be `has_fderiv_at_within.mono` and not a def